qmi/core/transport.py

```diff
@@ -1019,31 +1020,53 @@ def write(self, data: bytes) -> None:
         except vxi11.vxi11.Vxi11Exception as err:
             raise QMI_InstrumentException() from err
 
-    def read(self, nbytes: int, timeout: Optional[float]) -> bytes:
+    def read(self, nbytes: int, timeout: Optional[float] = None) -> bytes:
         self._check_is_open()
+        nbuf = len(self._read_buffer)
+        if nbuf >= nbytes:
+            # The requested number of bytes are already in the buffer. Return them immediately.
+            ret = self._read_buffer[:nbytes]
+            self._read_buffer = self._read_buffer[nbytes:]
+            return ret
+
         old_timeout = self._safe_instr.timeout
         if timeout:
             self._safe_instr.timeout = timeout
+
         try:
-            data = self._safe_instr.read_raw(nbytes)
+            while len(self._read_buffer) < nbytes:
```

Comment:
minor/major: only problem is that we're not really respecting our timeouts. if the batch is read faster than the timeout, we can basically read indefinitely.

Reply:
... up to `nbytes` indefinitely. So, not really an issue as we have a max size limit.